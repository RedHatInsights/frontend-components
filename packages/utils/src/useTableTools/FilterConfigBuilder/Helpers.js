export const stringToId = (string) => (
    string.split(' ').join('').toLowerCase()
);
