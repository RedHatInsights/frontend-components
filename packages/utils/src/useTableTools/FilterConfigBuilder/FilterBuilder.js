import { conditionalFilterType } from '@redhat-cloud-services/frontend-components/ConditionalFilter';
import { stringToId } from './Helpers';

class FilterBuilder {
    constructor(filterConfig) {
        this.filterConfig = filterConfig;
        this.config = this.filterConfig.config;
    }

    buildFilterFilterString = (configItem, value) => {
        const { type, filterString } = configItem;

        if (type !== 'hidden' && !value) { return []; }

        switch (type) {
            case conditionalFilterType.text:
                return [ filterString(value) ];
            case conditionalFilterType.checkbox:
                return value.map((filter) => (filterString(filter)));

            case conditionalFilterType.group:
                return filterString(value);

            case 'hidden':
                return filterString();

            default:
                return [];
        }
    }

    combineFilterStrings = (filterStringArray) => {
        const moreThanTwo = filterStringArray.map((f) => (f.length)).filter((fl) => (fl > 0)).length >= 2;
        return filterStringArray.map((fs) => (fs.join(' or '))).join(moreThanTwo ? ' and ' : '');
    }

    buildFilterString = (filters) => {
        const filterStringArray = this.config.map((configItem) => (
            this.buildFilterFilterString(configItem, filters[stringToId(configItem.label)])
        )).filter((f) => (f.length > 0));
        return this.combineFilterStrings(filterStringArray);
    }
}

export default FilterBuilder;
