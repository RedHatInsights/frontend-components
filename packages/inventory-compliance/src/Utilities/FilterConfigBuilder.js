import { conditionalFilterType } from '@redhat-cloud-services/frontend-components';
import FilterBuilder from './FilterBuilder';
import ChipBuilder from './ChipBuilder';

export const stringToId = (string) => (
    string.replace(' ', '').toLowerCase()
);

class FilterConfigBuilder {
    chipBuilder = null;
    filterBuilder = null;

    constructor(config) {
        this.config = config;
    }

    addConfigItem = (item) => (
        this.config = this.config.filter((i) => (
            i.label !== item.label
        )).concat(item)
    )

    getChipBuilder = () => (
        this.chipBuilder = this.chipBuilder ? this.chipBuilder : new ChipBuilder(this)
    )

    getFilterBuilder = () => (
        this.filterBuilder = this.filterBuilder ? this.filterBuilder : new FilterBuilder(this)
    )

    toTextFilterConfig = (item, handler, value) => ({
        type: conditionalFilterType.text,
        label: item.label,
        id: stringToId(item.label),
        filterValues: {
            value,
            [item.event || 'onChange']: (_event, selectedValues) => {
                handler(stringToId(item.label), selectedValues);
            }
        }
    });

    toCheckboxFilterConfig = (item, handler, value) => ({
        type: conditionalFilterType.checkbox,
        label: item.label,
        id: stringToId(item.label),
        filterValues: {
            value,
            items: item.items,
            onChange: (_event, selectedValues) => {
                handler(stringToId(item.label), selectedValues);
            }
        }
    });

    toRadioFilterConfig = (item, handler, value) => ({
        type: conditionalFilterType.radio,
        label: item.label,
        id: stringToId(item.label),
        filterValues: {
            value,
            items: item.items,
            onChange: (_event, selectedValues) => {
                handler(stringToId(item.label), selectedValues);
            }
        }
    });

    toFilterConfigItem = (item, handler, value) => {
        switch (item.type) {
            case conditionalFilterType.text:
                return this.toTextFilterConfig(item, handler, value);

            case conditionalFilterType.checkbox:
                return this.toCheckboxFilterConfig(item, handler, value);

            case conditionalFilterType.radio:
                return this.toRadioFilterConfig(item, handler, value);

            default:
                return null;
        }
    };

    buildConfiguration = (handler, states, props = {}) => ({
        ...props,
        items: this.config.map((item) => (
            this.toFilterConfigItem(item, handler, states[stringToId(item.label)])
        ))
    });

    defaultValueForFilter = (filter) => {
        switch (filter.type) {
            case conditionalFilterType.text:
                return '';
            case conditionalFilterType.checkbox:
                return [];
            default:
                return;
        }
    }

    initialDefaultState = (defaultStates = {}) => {
        let initialState = {};
        this.config.forEach((filter) => {
            const filterStateName = filter.label.replace(' ', '').toLowerCase();
            initialState[filterStateName] =
                defaultStates[filterStateName] || this.defaultValueForFilter(filter);
        });
        return initialState;
    }

    categoryLabelForValue = (value) => {
        const category = this.config.filter((category) => (
            category.items ? category.items.map((item) => item.value).includes(value) : false
        ))[0];

        return category ? category.label : value;
    };

    getCategoryForLabel = (query) => (
        this.config.filter((item) => (stringToId(item.label) === stringToId(query)))[0] || {}
    )

    getItemByLabelOrValue = (query, category) => {
        const items = this.getCategoryForLabel(category).items;
        const results = (items || []).filter((item) => (
            item.value === query || item.label === query
        ));

        if (results.length === 1) {
            return results[0];
        } else if (results.length > 1) {
            // eslint-disable-next-line no-console
            console.info(`Multiple items found for ${query} in ${category}! Returning first one.`);
            return results[0];
        } else {
            // eslint-disable-next-line no-console
            console.info('No item found for ' + query + ' in ', category);
        }
    }

    labelForValue = (value, category) => {
        const item = this.getItemByLabelOrValue(value, category);
        return item ? item.label : value;
    };

    valueForLabel = (label, category) => {
        const item = this.getItemByLabelOrValue(label, category);
        return item ? item.value : label;
    };

    applyFilterToObjectArray = (objects, activeFilters) => {
        let filteredObjects = [ ...objects ];
        Object.keys(activeFilters).forEach((filter) => {
            const category = this.getCategoryForLabel(filter);
            const value = activeFilters[filter];
            if (!category || !value) {
                return;
            }

            if (value.length > 0) {
                filteredObjects = category.filter(filteredObjects, value);
            }
        });

        return filteredObjects;
    }

    removeFilterFromFilterState = (currentState, filter) => (
        (typeof(currentState) === 'string') ? '' :
            currentState.filter((value) =>
                value !== filter
            )
    )

    removeFilterWithChip = (chips, activeFilters) => {
        const chipCategory = chips.category;
        const chipValue = this.valueForLabel(chips.chips[0].name, chipCategory);
        const stateProp = stringToId(chipCategory);
        const currentState = activeFilters[stateProp];
        const newFilterState = this.removeFilterFromFilterState(currentState, chipValue);

        return {
            ...activeFilters,
            [stateProp]: newFilterState
        };
    }
}

export default FilterConfigBuilder;
