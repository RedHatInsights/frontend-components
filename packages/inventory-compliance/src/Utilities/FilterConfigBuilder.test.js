import buildFilterConfig from './FilterBuilderConfigBuilder';
import FilterConfigBuilder from './FilterConfigBuilder';

describe('FilterConfigBuilder', () => {
    const config = buildFilterConfig({ selectedFilter: true, showPassFailFilter: true, policies: [] });
    let builder;

    beforeEach(() => {
        builder = new FilterConfigBuilder(config);
    });

    it('returns a filterConfig', () => {
        const states = {
            name: '',
            compliant: [],
            compliancescore: []
        };
        const builtConfig = builder.buildConfiguration(config, ()=> ({}), states);

        expect(builtConfig).toMatchSnapshot();
    });

    describe('categoryLabelForValue', () => {
        it('to return a matching category label', () => {
            expect(builder.categoryLabelForValue('true', 'compliant')).toMatchSnapshot();
            expect(builder.categoryLabelForValue('0-49', 'compliancescore')).toMatchSnapshot();
            expect(builder.categoryLabelForValue('Search term', 'name')).toMatchSnapshot();
        });
    });

    describe('labelForValue', () => {
        it('to return a matching label', () => {
            expect(builder.labelForValue('true', 'compliant')).toMatchSnapshot();
            expect(builder.labelForValue('0-49', 'compliancescore')).toMatchSnapshot();
            expect(builder.labelForValue('Search term', 'name')).toMatchSnapshot();
        });
    });

    describe('valueForLabel', () => {
        it('to return a matching value', () => {
            expect(builder.valueForLabel('Non-compliant', 'compliant')).toMatchSnapshot();
            expect(builder.valueForLabel('50 - 69%', 'compliancescore')).toMatchSnapshot();
            expect(builder.valueForLabel('Search term', 'name')).toMatchSnapshot();
        });
    });
});
