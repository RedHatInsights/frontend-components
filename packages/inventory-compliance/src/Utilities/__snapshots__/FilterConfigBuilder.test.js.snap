// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterConfigBuilder categoryLabelForValue to return a matching category label 1`] = `"true"`;

exports[`FilterConfigBuilder categoryLabelForValue to return a matching category label 2`] = `"0-49"`;

exports[`FilterConfigBuilder categoryLabelForValue to return a matching category label 3`] = `"Search term"`;

exports[`FilterConfigBuilder initialDefaultState to return a matching category label 1`] = `
Object {
  "filtergroup": undefined,
  "filterwithmultiplespaces": Array [],
  "name": "",
  "passed": Array [],
  "severity": Array [],
}
`;

exports[`FilterConfigBuilder labelForValue to return a matching label 1`] = `"true"`;

exports[`FilterConfigBuilder labelForValue to return a matching label 2`] = `"0-49"`;

exports[`FilterConfigBuilder labelForValue to return a matching label 3`] = `"Search term"`;

exports[`FilterConfigBuilder returns a filterConfig 1`] = `
Object {
  "compliancescore": Array [],
  "compliant": Array [],
  "filtergroup": Object {
    "1": Object {
      "2": true,
    },
  },
  "items": Array [
    Object {
      "filterValues": Object {
        "items": Array [
          Object {
            "label": "Show ",
            "value": "show",
          },
        ],
        "onChange": [Function],
        "value": undefined,
      },
      "id": "filterwithmultiplespaces",
      "label": "Filter with multiple spaces",
      "placeholder": "Filter by filter with multiple spaces",
      "type": "checkbox",
    },
    Object {
      "filterValues": Object {
        "groups": Array [
          Object {
            "items": Array [
              Object {
                "label": "Child 1",
                "type": "checkbox",
                "value": 1,
              },
              Object {
                "label": "Child 2",
                "type": "checkbox",
                "value": 2,
              },
            ],
            "label": "Parent 1",
            "type": "checkbox",
            "value": 1,
          },
          Object {
            "items": Array [
              Object {
                "label": "Parent 2 Child 1",
                "type": "checkbox",
                "value": 1,
              },
              Object {
                "label": "Parent 2 Child 2",
                "type": "checkbox",
                "value": 2,
              },
            ],
            "label": "Parent 2",
            "type": "checkbox",
            "value": 2,
          },
        ],
        "onChange": [Function],
        "selected": undefined,
      },
      "id": "filtergroup",
      "label": "Filter group",
      "type": "group",
    },
    Object {
      "filterValues": Object {
        "onChange": [Function],
        "value": "",
      },
      "id": "name",
      "label": "Name",
      "placeholder": "Filter by name",
      "type": "text",
    },
    Object {
      "filterValues": Object {
        "items": Array [
          Object {
            "label": <React.Fragment>
              <ExclamationCircleIcon
                className="ins-u-failed"
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
               High
            </React.Fragment>,
            "value": "high",
          },
          Object {
            "label": <React.Fragment>
              <ExclamationTriangleIcon
                className="ins-u-warning"
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
               Medium
            </React.Fragment>,
            "value": "medium",
          },
          Object {
            "label": <React.Fragment>
              <svg
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 18 18"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 0h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2zm6.67 10.46a1.67 1.67 0 1 0 0 3.338 1.67 1.67 0 0 0 0-3.338zm-1.586-6l.27 4.935a.435.435 0 0 0 .434.411H9.55c.232 0 .422-.18.435-.411l.27-4.936A.435.435 0 0 0 9.818 4h-2.3c-.25 0-.448.21-.435.46z"
                  fill="#3A9CA6"
                  fillRule="evenodd"
                />
              </svg>
               Low
            </React.Fragment>,
            "value": "low",
          },
          Object {
            "label": <React.Fragment>
              <QuestionCircleIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              />
               Unknown
            </React.Fragment>,
            "value": "unknown",
          },
        ],
        "onChange": [Function],
        "value": undefined,
      },
      "id": "severity",
      "label": "Severity",
      "placeholder": "Filter by severity",
      "type": "checkbox",
    },
    Object {
      "filterValues": Object {
        "items": Array [
          Object {
            "label": "Passed rules",
            "value": "passed",
          },
          Object {
            "label": "Failed rules",
            "value": "failed",
          },
        ],
        "onChange": [Function],
        "value": undefined,
      },
      "id": "passed",
      "label": "Passed",
      "placeholder": "Filter by passed",
      "type": "checkbox",
    },
  ],
  "name": "",
}
`;

exports[`FilterConfigBuilder valueForLabel to return a matching value 1`] = `"Non-compliant"`;

exports[`FilterConfigBuilder valueForLabel to return a matching value 2`] = `"50 - 69%"`;

exports[`FilterConfigBuilder valueForLabel to return a matching value 3`] = `"Search term"`;
